<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.54">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Sven Van Bael">
<meta name="dcterms.date" content="2024-10-07">

<title>Testing MS2PIP, a tool for peptide spectral library prediction, with C. elegans neuropeptides – Sven Van Bael</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">Sven Van Bael</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../Personal_projects/personal_projects.html"> 
<span class="menu-text">Personal projects</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../Academic_work/academic_work.html"> 
<span class="menu-text">Academic work</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#c.-elegans-neuropeptide-spectral-library-prediction-with-ms2pip" id="toc-c.-elegans-neuropeptide-spectral-library-prediction-with-ms2pip" class="nav-link" data-scroll-target="#c.-elegans-neuropeptide-spectral-library-prediction-with-ms2pip"><span class="header-section-number">2</span> <em>C. elegans</em> neuropeptide spectral library prediction with MS2PIP</a></li>
  <li><a href="#results" id="toc-results" class="nav-link" data-scroll-target="#results"><span class="header-section-number">3</span> Results</a>
  <ul class="collapse">
  <li><a href="#packages-used" id="toc-packages-used" class="nav-link" data-scroll-target="#packages-used"><span class="header-section-number">3.1</span> Packages used</a></li>
  <li><a href="#peak-picking-correction" id="toc-peak-picking-correction" class="nav-link" data-scroll-target="#peak-picking-correction"><span class="header-section-number">3.2</span> Peak picking correction</a></li>
  <li><a href="#comparing-the-ions-dotp-values-for-the-different-libraries" id="toc-comparing-the-ions-dotp-values-for-the-different-libraries" class="nav-link" data-scroll-target="#comparing-the-ions-dotp-values-for-the-different-libraries"><span class="header-section-number">3.3</span> Comparing the ions’ dotp values for the different libraries</a></li>
  <li><a href="#library-prediction-is-negatively-affected-by-post-translational-modifications-and-higher-charge-states" id="toc-library-prediction-is-negatively-affected-by-post-translational-modifications-and-higher-charge-states" class="nav-link" data-scroll-target="#library-prediction-is-negatively-affected-by-post-translational-modifications-and-higher-charge-states"><span class="header-section-number">3.4</span> Library prediction is negatively affected by post-translational modifications and higher charge states</a></li>
  </ul></li>
  <li><a href="#conclusions" id="toc-conclusions" class="nav-link" data-scroll-target="#conclusions"><span class="header-section-number">4</span> Conclusions</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Testing MS2PIP, a tool for peptide spectral library prediction, with <em>C. elegans</em> neuropeptides</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Sven Van Bael </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">October 7, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>In mass spectrometry-based proteomics and peptidomics, peptide identification for targeted (quantitative) techniques, such as parallel reaction monitoring (PRM) and data-independent acquisition (DIA), rely on matching the obtained fragmentation patterns with a reference fragmentation pattern of the peptide(s) of interest. Although the individual masses of the (b and y) fragment ions can be easily calculated, their (relative) intensities are far more challenging to accurately predict, as they are influenced by amino acid composition, location within the peptide, and overall charge state. Hence, most PRM or DIA techniques will employ empirically collected LC-MS/MS fragmentation data of the peptides of interest, commonly referred to as a “spectral library”. A major downside here is that you are depending on the peptide being observed in previously acquired LC-MS/MS runs, be it from running biological reference samples, or from synthetic versions of your peptide(s) of interest.</p>
<p>Generating a peptide library is therefore no trivial task, often requiring multiple LC-MS/MS runs, filtering and selection of high-quality ions. Over the years, several research groups undertook the challenge to train deep learning models (trained on real-life proteomics data) that would be able to generate such libraries from scratch, only requiring the peptide sequence and charge (which can also be predicted fairly accurately). The two most promising models are MS2PIP from the CompOmics group in Gent <span class="citation" data-cites="gabriels2019">(<a href="#ref-gabriels2019" role="doc-biblioref">Gabriels et al., 2019</a>)</span>, and PROSIT from the Proteomics Bioinformatics group in Munich <span class="citation" data-cites="gessulat2019">(<a href="#ref-gessulat2019" role="doc-biblioref">Gessulat et al., 2019</a>)</span>.</p>
<p>During my PhD and postdoc, I’ve been working on the identification of neuropeptides in the model organism <em>C. elegans</em>, and the development of targeted LC-MS/MS approaches for their relative quantification. Although they are both called “peptides”, neuropeptides and the typical tryptic peptides targeted in proteomics methods differ enough to necessitate tailored LC-MS-MS approaches. For one, neuropeptides are typically shorter in length, and do not have the signature C-terminal basic residue as is seen in tryptic peptides (peptidomics workflows do not include a digestion step, since the peptides are already produced within the organism). Hence, models that are trained on tryptic peptides for the prediction of fragmentation spectra produce disappointing results when you feed them neuropeptide sequences. Another problem are the post-translational modifications that are observed on neuropeptides, of which C-terminal amidation and N-terminal pyroglutamation are the most common in neuropeptides. Models trained on tryptic peptides do not factor in these modifications, as they are not common in proteomics. On the other hand, PROSIT <strong>always</strong> assumes that the peptides have a carbamidomethyl modification on cysteine, as proteomics (but not peptidomics!) sample preparation incorporates a cysteine alkylation step to block Cys-Cys bond formation.</p>
<p>Here, I have tested how two MS2PIP models (immunoHCD and HCDch2) and compared them to my own empirically collected neuropeptide spectral library. The choice for MS2PIP was made based on the fact that (1) they have a model trained on non-tryptic peptides (immunoHCD), (2) that they allow user-specified post-translational modifications, and (3) unlike PROSIT, MS2PIP does not have the built-in assumption that all peptides have carbamidomethylated cysteines, unless specified otherwise.</p>
</section>
<section id="c.-elegans-neuropeptide-spectral-library-prediction-with-ms2pip" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="c.-elegans-neuropeptide-spectral-library-prediction-with-ms2pip"><span class="header-section-number">2</span> <em>C. elegans</em> neuropeptide spectral library prediction with MS2PIP</h2>
<p><em>While I provided major contributions in the generation of all the data that is described here, it is in its current (unpublished) form not solely mine to share, and is still property of the <a href="https://bio.kuleuven.be/df/lt">Temmerman lab</a>. Hence, all references to neuropeptide IDs and specifics regarding the ions targeted have been removed from the datasets.</em></p>
<p><em>This analysis was originally performed medio 2023. Any specifics regarding MS2PIP, the available models, and user interface might have changed over time.</em></p>
<p><a href="https://iomics.ugent.be/ms2pip/">MS2PIP</a> offers a variety of models for spectral library prediction (<a href="#fig-MS2PIP_models" class="quarto-xref">Figure&nbsp;1</a>). Only one model, immunoHCD, is trained on (non-tryptic) immunopeptides and is the model of interest to be tested here. As a reference, I included HCDch2, which is trained on peptides originating from tryptic digests. The expectation being that immunoHCD should outperform HCDch2, since neuropeptides are probably more similar to immunopeptides than to tryptic peptides.</p>
<div id="fig-MS2PIP_models" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-MS2PIP_models-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="MS2PIP_models.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-MS2PIP_models-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Overview of all models available in MS2PIP. Highlighted are the models that are compatible with my data: trained on data collected with HCD fragmentation and without labelling. A choice was made for the immunoHCD model (because it is trained on non-tryptic immunopeptides) and HCDch2 (which is trained on tryptic peptides) as a reference for comparison.
</figcaption>
</figure>
</div>
<p>As specified in the MS2PIP “How To”, a <code>PEPREC</code> input file containing peptide IDs, peptide sequences, modifications, and charge was created, containing the ions of 300 individual neuropeptides (the same amount as in my empirically collected spectral library). User-defined post-translational modifications and their mass shifts are specified in the <code>config</code> file, and consist of:</p>
<ul>
<li>Oxidation (Methionine): +15.994915;</li>
<li>Amidation (C-terminus): -0984016;</li>
<li>Pyroglutamation (Glutamine): -17.026549;</li>
<li>Pyroglutamation (Glutamic acid): -18.010565;</li>
<li>Half of disulfide bridge (Cysteine): -1.007825;</li>
</ul>
<p>All spectral libraries (my own empirical one, and the MS2PIP generated immunoHCD and HCDch2) are individually used in a <a href="https://skyline.ms/project/home/software/skyline/begin.view">Skyline</a> workflow for the analysis of a high-quality PRM dataset. This PRM data was recorded on a sample pool of synthetic <em>C. elegans</em> neuropeptides, and has been analyzed in great detail before, so I know that the spectra of all 300 neuropeptides are present in this dataset and can be detected and quantified when using my own empirical spectral library.</p>
</section>
<section id="results" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="results"><span class="header-section-number">3</span> Results</h2>
<section id="packages-used" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="packages-used"><span class="header-section-number">3.1</span> Packages used</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggpubr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggbeeswarm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="peak-picking-correction" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="peak-picking-correction"><span class="header-section-number">3.2</span> Peak picking correction</h3>
<p>During the analysis , Skyline’s peak picking algorithm goes through the PRM data, and based on similarities with what is in the spectral library, will automatically add integration borders for MS1 and MS2 peaks. These integration borders are prone to shift a bit between Skyline analyses, especially when using different libraries. So for a fair comparison, I made sure that (approximately) the same region of integration is selected between the three different spectral libraries. In <a href="#fig-peak_pick_noncorr" class="quarto-xref">Figure&nbsp;2</a>, I plotted the MS1 area under the curve of individual ions when using the empirical database versus those when using immunoHCD and HCDch2 libraries. Immediately, you can see that some ions have vastly different AUC values depending on when the empirical or the immunoHCD/HCDch2 was used. Since I’m using my empirical library as a control, I will manually correct the integration borders in the immunoHCD and HCDch2 analyses to correspond with those where the empirical spectral library was used.</p>
<div class="cell">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Reading in the csv file containing the different AUCs of individual peptides, analyzed with the three different spectral library.</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>df_noncorr <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"Empirical-immunoHCD-HCDch2.csv"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot AUCs of data analyzed with the empirical versus immunoHCD library.</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>Emp_immuno <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_noncorr) <span class="sc">+</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">Total Area MS1_Empirical</span><span class="st">`</span>,</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">y =</span> <span class="st">`</span><span class="at">Total Area MS1_immunoHCD</span><span class="st">`</span>,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>                 <span class="at">color =</span> (<span class="st">`</span><span class="at">Total Area MS1_Empirical</span><span class="st">`</span><span class="sc">/</span><span class="st">`</span><span class="at">Total Area MS1_immunoHCD</span><span class="st">`</span>) <span class="sc">&lt;</span> <span class="fl">0.95</span> <span class="sc">|</span> (<span class="st">`</span><span class="at">Total Area MS1_Empirical</span><span class="st">`</span><span class="sc">/</span><span class="st">`</span><span class="at">Total Area MS1_immunoHCD</span><span class="st">`</span>) <span class="sc">&gt;</span> <span class="fl">1.3</span>,</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>                 <span class="at">fill =</span> (<span class="st">`</span><span class="at">Total Area MS1_Empirical</span><span class="st">`</span><span class="sc">/</span><span class="st">`</span><span class="at">Total Area MS1_immunoHCD</span><span class="st">`</span>) <span class="sc">&lt;</span> <span class="fl">0.95</span> <span class="sc">|</span> (<span class="st">`</span><span class="at">Total Area MS1_Empirical</span><span class="st">`</span><span class="sc">/</span><span class="st">`</span><span class="at">Total Area MS1_immunoHCD</span><span class="st">`</span>) <span class="sc">&gt;</span> <span class="fl">1.3</span>), <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">3</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">name =</span> <span class="st">"Empirical"</span>, <span class="at">trans =</span> <span class="st">"log10"</span>) <span class="sc">+</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">name =</span> <span class="st">"immunoHCD"</span>, <span class="at">trans =</span> <span class="st">"log10"</span>) <span class="sc">+</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#000000"</span>, <span class="st">"#A50F15"</span>)) <span class="sc">+</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#0000004D"</span>, <span class="st">"#A50F154D"</span>)) <span class="sc">+</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>),</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>, <span class="at">color =</span> <span class="st">"black"</span>))</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot AUCs of data analyzed with the empirical versus HCDch2 library.</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>Emp_HCDch2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_noncorr) <span class="sc">+</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">Total Area MS1_Empirical</span><span class="st">`</span>,</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>                 <span class="at">y =</span> <span class="st">`</span><span class="at">Total Area MS1_HCDch2</span><span class="st">`</span>,</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>                 <span class="at">color =</span> (<span class="st">`</span><span class="at">Total Area MS1_Empirical</span><span class="st">`</span><span class="sc">/</span><span class="st">`</span><span class="at">Total Area MS1_HCDch2</span><span class="st">`</span>) <span class="sc">&lt;</span> <span class="fl">0.95</span> <span class="sc">|</span> (<span class="st">`</span><span class="at">Total Area MS1_Empirical</span><span class="st">`</span><span class="sc">/</span><span class="st">`</span><span class="at">Total Area MS1_HCDch2</span><span class="st">`</span>) <span class="sc">&gt;</span> <span class="fl">1.3</span>,</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>                 <span class="at">fill =</span> (<span class="st">`</span><span class="at">Total Area MS1_Empirical</span><span class="st">`</span><span class="sc">/</span><span class="st">`</span><span class="at">Total Area MS1_HCDch2</span><span class="st">`</span>) <span class="sc">&lt;</span> <span class="fl">0.95</span> <span class="sc">|</span> (<span class="st">`</span><span class="at">Total Area MS1_Empirical</span><span class="st">`</span><span class="sc">/</span><span class="st">`</span><span class="at">Total Area MS1_HCDch2</span><span class="st">`</span>) <span class="sc">&gt;</span> <span class="fl">1.3</span>), <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">3</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">name =</span> <span class="st">"Empirical"</span>, <span class="at">trans =</span> <span class="st">"log10"</span>) <span class="sc">+</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">name =</span> <span class="st">"HCDch2"</span>, <span class="at">trans =</span> <span class="st">"log10"</span>) <span class="sc">+</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#000000"</span>, <span class="st">"#A50F15"</span>)) <span class="sc">+</span></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#0000004D"</span>, <span class="st">"#A50F154D"</span>)) <span class="sc">+</span></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>),</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>, <span class="at">color =</span> <span class="st">"black"</span>))</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a><span class="co">#Arrange plots.</span></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a><span class="fu">ggarrange</span>(Emp_immuno, Emp_HCDch2, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>), <span class="at">font.label =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">20</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-peak_pick_noncorr" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-peak_pick_noncorr-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="Testing_MS2PIP_files/figure-html/fig-peak_pick_noncorr-1.png" class="img-fluid figure-img" width="787">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peak_pick_noncorr-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: MS1 areas under the curve of individual identified neuropeptide ions in the <em>C. elegans</em> neuropeptide PRM dataset, all peak-picked automatically by the Skyline algorithm. <strong>(A)</strong> shows AUCs from the PRM data analyzed with the empirical versus immunoHCD spectral library, and <strong>(B)</strong> shows the Empirical versus HCDch2 spectral library. Both axes are on a log<sub>10</sub> scale. Red dots are ions where the ratio of the two AUCs <span class="math inline">\((\frac{AUC_{Empirical}}{AUC_{MS2PIP}})\)</span> is lower than 0.95 or higher than 1.3.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Based on the information in <a href="#fig-peak_pick_noncorr" class="quarto-xref">Figure&nbsp;2</a>, I manually corrected the integration windows of 15 ions in the analysis with the immunoHCD library, and 32 ions in the analysis with the HCDch2 library. Going by those numbers, it already looks that the HCDch2 is not performing very well, as peak picking seems to fail more often when using that library. Remaking the plots of <a href="#fig-peak_pick_noncorr" class="quarto-xref">Figure&nbsp;2</a> after manual correction, it is clear that the integration windows between the three analyses are similar enough to make a fair comparison possible (<a href="#fig-peak_pick_corr" class="quarto-xref">Figure&nbsp;3</a>).</p>
<div class="cell">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Reading in the csv file containing the different AUCs of individual peptides, analyzed with the three different spectral library - MS1 integration windows corrected.</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>df_corr <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"Empirical-immunoHCD-HCDch2_MS1_corrected.csv"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot AUCs of data analyzed with the empirical versus immunoHCD library.</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>Emp_immuno_corr <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_corr) <span class="sc">+</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">Total Area MS1_Empirical</span><span class="st">`</span>,</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">y =</span> <span class="st">`</span><span class="at">Total Area MS1_immunoHCD</span><span class="st">`</span>),</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>             <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">fill =</span> <span class="st">"#0000004D"</span>, <span class="at">size =</span> <span class="dv">3</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">name =</span> <span class="st">"Empirical"</span>, <span class="at">trans =</span> <span class="st">"log10"</span>) <span class="sc">+</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">name =</span> <span class="st">"immunoHCD"</span>, <span class="at">trans =</span> <span class="st">"log10"</span>) <span class="sc">+</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>),</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>, <span class="at">color =</span> <span class="st">"black"</span>))</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot AUCs of data analyzed with the empirical versus HCDch2 library.</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>Emp_HCDch2_corr <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df_corr) <span class="sc">+</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">Total Area MS1_Empirical</span><span class="st">`</span>,</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>                 <span class="at">y =</span> <span class="st">`</span><span class="at">Total Area MS1_HCDch2</span><span class="st">`</span>),</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>             <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">fill =</span> <span class="st">"#0000004D"</span>, <span class="at">size =</span> <span class="dv">3</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">name =</span> <span class="st">"Empirical"</span>, <span class="at">trans =</span> <span class="st">"log10"</span>) <span class="sc">+</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">name =</span> <span class="st">"HCDch2"</span>, <span class="at">trans =</span> <span class="st">"log10"</span>) <span class="sc">+</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>),</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>, <span class="at">color =</span> <span class="st">"black"</span>))</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="co">#Arrange plots.</span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a><span class="fu">ggarrange</span>(Emp_immuno_corr, Emp_HCDch2_corr, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>), <span class="at">font.label =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">20</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-peak_pick_corr" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-peak_pick_corr-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="Testing_MS2PIP_files/figure-html/fig-peak_pick_corr-1.png" class="img-fluid figure-img" width="787">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-peak_pick_corr-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: MS1 areas under the curve of individual identified neuropeptide ions in the <em>C. elegans</em> neuropeptide PRM dataset, manually corrected after automatic peak picking by the Skyline algorithm. <strong>(A)</strong> shows AUCs from the PRM data analyzed with the empirical versus immunoHCD spectral library, and <strong>(B)</strong> shows the Empirical versus HCDch2 spectral library. Both axes are on a log<sub>10</sub> scale.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="comparing-the-ions-dotp-values-for-the-different-libraries" class="level3" data-number="3.3">
<h3 data-number="3.3" class="anchored" data-anchor-id="comparing-the-ions-dotp-values-for-the-different-libraries"><span class="header-section-number">3.3</span> Comparing the ions’ dotp values for the different libraries</h3>
<p>For each identified ion, Skyline provides a metric that expresses how well the identified fragmentation peaks in your data correspond to the reference spectrum that is present in the spectral library. This metric, <a href="https://skyline.ms/announcements/home/support/thread.view?rowid=20003">the dotp (or normalized contrast angle)</a>, ranges from 0 (no similarity at al) to 1 (perfect similarity, observed spectrum and library spectrum are identical). Since the PRM dataset that I am analyzing are pure synthetic neuropeptides, I know that these fragmentation patterns will be of high quality, and each one will correspond to a neuropeptide ion (as these are the only things present in the sample, no background). Hence, I can use the dotp as a quality indicator of the ion fragmentation spectra in the library (which is actually the exact opposite of what the metric is normally used for, and only possible because my PRM data contains purely my neuropeptides of interest).</p>
<p>Hypothetically, when the neuropeptde PRM data is analyzed with a high-quality spectral library, the bulk of the dotp values will be high, almost 1. On the other hand, when the spectral library is of low quality, or the library fragment ions do not correspond with what is in the PRM dataset, the dotp values will drop. In <a href="#fig-dotp" class="quarto-xref">Figure&nbsp;4</a>, it is obvious that the dotp values of both predicted spectral libraries are far off when compared to the empirical library. Even though the immunoHCD model produces a slightly better library when compared to HCDch2, a significant amount of ions are below 0.75, which is often used as a cut-off for the quality of an identification.</p>
<div class="cell">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Transform dataframe containing dotp values into long format.</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>df_corr_long <span class="ot">&lt;-</span> df_corr <span class="sc">%&gt;%</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(<span class="dv">4</span><span class="sc">:</span><span class="dv">12</span>), <span class="at">names_sep =</span> <span class="st">"_"</span>, <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"Metric"</span>, <span class="st">"Library"</span>), <span class="at">values_to =</span> <span class="st">"Value"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> Metric, <span class="at">values_from =</span> Value)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>df_corr_long<span class="sc">$</span>Library <span class="ot">&lt;-</span> <span class="fu">factor</span>(df_corr_long<span class="sc">$</span>Library, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Empirical"</span>, <span class="st">"immunoHCD"</span>, <span class="st">"HCDch2"</span>))</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="co">#Plot dotp values for all ions, per library used.</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_corr_long) <span class="sc">+</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_quasirandom</span>(<span class="fu">aes</span>(<span class="at">x =</span> Library, <span class="at">y =</span> <span class="st">`</span><span class="at">Library Dot Product</span><span class="st">`</span>, <span class="at">color =</span> Library, <span class="at">fill =</span> Library), <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">size =</span> <span class="dv">3</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">0.75</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"#5A5A5A"</span>) <span class="sc">+</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#FEA82F"</span>, <span class="st">"#4D2D52"</span>, <span class="st">"#003f5c"</span>)) <span class="sc">+</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#FEA82F4D"</span>, <span class="st">"#4D2D524D"</span>, <span class="st">"#003f5c4D"</span>)) <span class="sc">+</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>),</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">16</span>, <span class="at">color =</span> <span class="st">"black"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-dotp" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-dotp-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="Testing_MS2PIP_files/figure-html/fig-dotp-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-dotp-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Dotp values for the PRM dataset analyzed with the three different spectral libraries. Grey line denotes the quality cut-off at 0.75.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="library-prediction-is-negatively-affected-by-post-translational-modifications-and-higher-charge-states" class="level3" data-number="3.4">
<h3 data-number="3.4" class="anchored" data-anchor-id="library-prediction-is-negatively-affected-by-post-translational-modifications-and-higher-charge-states"><span class="header-section-number">3.4</span> Library prediction is negatively affected by post-translational modifications and higher charge states</h3>
<p>C-terminal amidation and N-terminal pyroglutamation are two highly common post-translational modifications that are quite common in neuropeptides, but are generally not taken into account for proteomics purposes. Interestingly, these two modifications do happen to have a large impact on the overall charge state of the peptide at acidic pH (at which most peptide-based LC-MS/MS analyses are performed).</p>
<p>Due to its relatively mild basicity, the proton on the peptide’s N-terminal amine is mobile, and can move around along the amide bonds within the peptide <span class="citation" data-cites="kinter2000a">(<a href="#ref-kinter2000a" role="doc-biblioref">Kinter &amp; Sherman, 2000</a>)</span>. One can hypothesize that N-terminal pyroglutamation severely diminishes this proton mobility, as the significantly more basic pyrrole ring will lock the proton firmly in it’s place (<a href="#fig-PTMs" class="quarto-xref">Figure&nbsp;5</a>). On the other hand, the peptide C-terminal normally carries no charge at acidic pH. However, the resulting amide group that occurs after C-terminal amidation will likely be positively charged at an acidic pH (<a href="#fig-PTMs" class="quarto-xref">Figure&nbsp;5</a>).</p>
<div id="fig-PTMs" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-PTMs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="PTMs.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-PTMs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: The effect of N-terminal pyroglutamation and C-terminal amidation on overall peptide charge. The pyrrole ring that is formed in the pyroglutamation reaction locks a proton that otherwise can move around along the peptide amide bonds, while amidation modifies the normally neutral C-terminus into a positively charged one.
</figcaption>
</figure>
</div>
<p>It is fair to assume that both these modifications have a severe impact on peptide charge and fragmentation pattern, for which the currently available prediction models cannot accommodate. To find out if this is also the case for the <em>C. elegans</em> neuropeptides, I’ve plotted the dotp values of the different neuropeptide ions according to their different modifications: amidation, pyroglutamation, amidation &amp; pyroglutamation, or none at all (<a href="#fig-dotp_charge_ptm" class="quarto-xref">Figure&nbsp;6</a>).</p>
<div class="cell">
<details class="code-fold">
<summary>Show code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Set factors.</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>df_corr_long<span class="sc">$</span><span class="st">`</span><span class="at">Precursor Charge</span><span class="st">`</span> <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(df_corr_long<span class="sc">$</span><span class="st">`</span><span class="at">Precursor Charge</span><span class="st">`</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>df_corr_long<span class="sc">$</span><span class="st">`</span><span class="at">PTM</span><span class="st">`</span> <span class="ot">&lt;-</span> <span class="fu">factor</span>(df_corr_long<span class="sc">$</span><span class="st">`</span><span class="at">PTM</span><span class="st">`</span>, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Pyroglutamation"</span>, <span class="st">"Amidation"</span>, <span class="st">"Pyroglutamation,</span><span class="sc">\n</span><span class="st">Amidation"</span>, <span class="st">"None"</span>))</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Create plot split by charge and PTM.</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_corr_long) <span class="sc">+</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(.<span class="sc">~</span>PTM) <span class="sc">+</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">Precursor Charge</span><span class="st">`</span>, <span class="at">y =</span> <span class="st">`</span><span class="at">Library Dot Product</span><span class="st">`</span>, <span class="at">color =</span> Library, <span class="at">fill =</span> Library)) <span class="sc">+</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">0.75</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"#5A5A5A"</span>) <span class="sc">+</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#FEA82F"</span>, <span class="st">"#4D2D52"</span>, <span class="st">"#003f5c"</span>)) <span class="sc">+</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#FEA82F4D"</span>, <span class="st">"#4D2D524D"</span>, <span class="st">"#003f5c4D"</span>)) <span class="sc">+</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text =</span> <span class="fu">element_text</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="dv">16</span>),</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>),</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>),</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"bottom"</span>, </span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">14</span>),</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">"#B4B4B4"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-dotp_charge_ptm" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-dotp_charge_ptm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="Testing_MS2PIP_files/figure-html/fig-dotp_charge_ptm-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-dotp_charge_ptm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: Dotp values for the PRM dataset analyzed with the three different spectral libraries, split for precursor charge and post-translational modification
</figcaption>
</figure>
</div>
</div>
</div>
<p>Pyroglutamation seems to have the most detrimental effect on fragment spectra prediction, with almost no ions having a dotp over 0.75 (<a href="#fig-dotp_charge_ptm" class="quarto-xref">Figure&nbsp;6</a>, first panel). For amidated neuropeptides, the model’s predictive success seems to depend on precursor charge, where dotp values drops for amidated neuropeptides with higher charge states (<a href="#fig-dotp_charge_ptm" class="quarto-xref">Figure&nbsp;6</a>, second panel). This could be a consequence of the amidation modification resulting in an additional charge (<a href="#fig-PTMs" class="quarto-xref">Figure&nbsp;5</a>). A combination of these both effects is seen for neuropeptides that are pyroglutamated and amidated, where most dotp values are below 0.75, and drop further with higher charges (<a href="#fig-dotp_charge_ptm" class="quarto-xref">Figure&nbsp;6</a>, third panel). Although the immunoHCD model performs best for neuropeptides with no modifications at all, it still cannot match the results obtained with an empirical spectral library (<a href="#fig-dotp_charge_ptm" class="quarto-xref">Figure&nbsp;6</a>, last panel). As expected, the HCDch2 model, which is trained on tryptic peptides, performs the worst of all in almost all instances.</p>
</section>
</section>
<section id="conclusions" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="conclusions"><span class="header-section-number">4</span> Conclusions</h2>
<p>Tools such as MS2PIP are an amazing addition for predicting peptide fragmentation patterns in proteomics workflows. However, for more specialized mass spectrometry-based research that does not depend on tryptic peptides, predictions are often severely flawed. The MS2PIP team partially tried to solve this by training new models on data containing non-tryptic peptides, such as immunopeptides.</p>
<p>It seems that just broadening the training data to non-tryptic peptides is not enough, since some “unusual” modifications present on (<em>C. elegans</em>) neuropeptides are enough to result in unreliable predictions from the models (<a href="#fig-dotp_charge_ptm" class="quarto-xref">Figure&nbsp;6</a>). Hopefully, future MS2PIP models will include more peptides originating from various sources beyond tryptic digests and immunopeptides, and have a higher variety of post-translational modifications.</p>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list">
<div id="ref-gabriels2019" class="csl-entry" role="listitem">
Gabriels, R., Martens, L., &amp; Degroeve, S. (2019). Updated MS²PIP web server delivers fast and accurate MS² peak intensity prediction for multiple fragmentation methods, instruments and labeling techniques. <em>Nucleic Acids Research</em>, <em>47</em>(W1), W295–W299. <a href="https://doi.org/10.1093/nar/gkz299">https://doi.org/10.1093/nar/gkz299</a>
</div>
<div id="ref-gessulat2019" class="csl-entry" role="listitem">
Gessulat, S., Schmidt, T., Zolg, D. P., Samaras, P., Schnatbaum, K., Zerweck, J., Knaute, T., Rechenberger, J., Delanghe, B., Huhmer, A., Reimer, U., Ehrlich, H.-C., Aiche, S., Kuster, B., &amp; Wilhelm, M. (2019). Prosit: proteome-wide prediction of peptide tandem mass spectra by deep learning. <em>Nature Methods</em>, <em>16</em>(6), 509–518. <a href="https://doi.org/10.1038/s41592-019-0426-7">https://doi.org/10.1038/s41592-019-0426-7</a>
</div>
<div id="ref-kinter2000a" class="csl-entry" role="listitem">
Kinter, M., &amp; Sherman, N. E. (2000). <em>Protein Sequencing and Identification Using Tandem Mass Spectrometry</em>. <a href="https://doi.org/10.1002/0471721980">https://doi.org/10.1002/0471721980</a>
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>